### PR TITLE
pd upgrade refactor

### PR DESCRIPTION
**Pull request by xiaojingchen**
this pr just abstracting the pd upgrader function into Upgrader interface, the changes is also to prepare for the feature of graceful upgrade feature. the tikv/tidb's  upgrade refactor will be committed by another pr

@weekface @tennix @onlymellb @gregwebs  PTAL

### CODE REVIEW

File: pkg/manager/member/utils.go [hunk 1]
```diff
@@ -156,8 +156,8 @@ func SetServiceLastAppliedConfigAnnotation(svc *corev1.Service) error {
 	return nil
 }
 
-// EqualService compare the new Service's spec with old Service's last applied config
-func EqualService(new, old *corev1.Service) (bool, error) {
+// equalService compare the new Service's spec with old Service's last applied config
+func equalService(new, old *corev1.Service) (bool, error) {
```
Comment (weekface) [inline pkg/manager/member/utils.go]: `serviceDeepEqual` may be a better function name.

File: pkg/manager/member/utils.go [hunk 2]
```diff
@@ -156,8 +156,8 @@ func SetServiceLastAppliedConfigAnnotation(svc *corev1.Service) error {
 	return nil
 }
 
-// EqualService compare the new Service's spec with old Service's last applied config
-func EqualService(new, old *corev1.Service) (bool, error) {
+// equalService compare the new Service's spec with old Service's last applied config
```
Comment (weekface) [inline pkg/manager/member/utils.go]: compare -> compares

File: pkg/manager/member/pd_member_manager.go [hunk 3]
```diff
@@ -180,8 +183,10 @@ func (pmm *pdMemberManager) syncPDStatefulSetForTidbCluster(tc *v1alpha1.TidbClu
 		return err
 	}
 
-	if err = pmm.upgrade(tc, oldPDSet, newPDSet); err != nil {
-		return err
+	if !equalTemplate(newPDSet.Spec.Template, oldPDSet.Spec.Template) {
+		if err = pmm.pdUpgrader.Upgrade(tc, oldPDSet, newPDSet); err != nil {
```
Comment (weekface) [inline pkg/manager/member/pd_member_manager.go]: `err = ` -> `err := `

File: pkg/manager/member/pd_member_manager.go [hunk 4]
```diff
@@ -196,11 +201,7 @@ func (pmm *pdMemberManager) syncPDStatefulSetForTidbCluster(tc *v1alpha1.TidbClu
 		}
 	}
 
-	euqal, err := EqualStatefulSet(*newPDSet, *oldPDSet)
-	if err != nil {
-		return err
-	}
-	if !euqal {
+	if !equalStatefulSet(*newPDSet, *oldPDSet) {
```
Comment (weekface) [inline pkg/manager/member/pd_member_manager.go]: Why use `*newPDSet ` instead of `newPDSet `?

File: pkg/manager/member/pd_upgrader.go [hunk 5]
```diff
@@ -0,0 +1,30 @@
+package member
```
Comment (weekface) [inline pkg/manager/member/pd_upgrader.go]: Add License header

File: pkg/manager/member/pd_upgrader.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+import (
+	"github.com/pingcap/tidb-operator/pkg/apis/pingcap.com/v1alpha1"
+	apps "k8s.io/api/apps/v1beta1"
+)
+
+type pdUpgrader struct{}
+
+// NewPDUpgrader return a pdUpgrader
```
Comment (weekface) [inline pkg/manager/member/pd_upgrader.go]: `return ` -> `returns `

File: pkg/manager/member/pd_upgrader.go [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+func (pu *pdUpgrader) Upgrade(tc *v1alpha1.TidbCluster, oldSet *apps.StatefulSet, newSet *apps.StatefulSet) error {
+	tc.Status.PD.Phase = v1alpha1.UpgradePhase
+	return nil
+}
+
+type fakePDUpgrader struct{}
+
+// NewFakePDUpgrader return a fakePDUpgrader
```
Comment (weekface) [inline pkg/manager/member/pd_upgrader.go]: return -> returns

File: pkg/manager/member/pd_upgrader_test.go [hunk 8]
```diff
@@ -0,0 +1,73 @@
+package member
```
Comment (weekface) [inline pkg/manager/member/pd_upgrader_test.go]: Add License header

File: pkg/manager/member/upgrader.go [hunk 9]
```diff
@@ -0,0 +1,12 @@
+package member
```
Comment (weekface) [inline pkg/manager/member/upgrader.go]: license

File: pkg/manager/member/upgrader.go [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+
+import (
+	"github.com/pingcap/tidb-operator/pkg/apis/pingcap.com/v1alpha1"
+	apps "k8s.io/api/apps/v1beta1"
+)
+
+// Upgrader implements the logic for upgrade the tidb cluster.
```
Comment (weekface) [inline pkg/manager/member/upgrader.go]: `for upgrade ` -> `for upgrading`

**Comment by tennix** [comment]
Can we add `needUpgrade` to `Upgrader`?

**Comment by xiaojingchen** [comment]
nice, the pd's ```needUpgrade``` function had been deleted in this pr, the other ```needUpgrade``` functions will be deleted in next pr.

**Pull request merged**